packages/stdlib/src/ring1/async.js

```diff
@@ -0,0 +1,49 @@
+// @ts-check
+import { makePromise } from '../ring0/main';
+
+/**
+ * Resolve promise with value when body returns falsy.
+ *
+ * @template T
+ * @param {() => T | Promise<T>} body perform side-effects, and return truthy if
+ * we should run again
+ * @returns {Promise<void>}
+ */
+const asyncDoWhile = body => {
+  const loop = async (resolve, reject) => {
+    const doContinue = await body();
+    if (!doContinue) {
+      // Resolve the outermost promise.
+      resolve(undefined);
+      return;
+    }
+
+    // Do the loop again.  We are careful not to await so that we don't create a
+    // promise chain.
+    loop(resolve, reject).catch(reject);
+  };
+
+  return makePromise((resolve, reject) => loop(resolve, reject).catch(reject));
+};
+harden(asyncDoWhile);
+export { asyncDoWhile };
+
+/**
+ * Return an async iterable that produces iterator results via next.
+ *
+ * @template T,TReturn
+ * @param {() => IteratorResult<T, TReturn> | Promise<IteratorResult<T,
+ * TReturn>>} next produce an iterator result
+ * @returns {AsyncIterable<T>}
+ */
+const asyncGenerate = next => {
```

Comment:
I tried to like this but I don't. It is at a weird intermediate level of abstraction. Most of the work is still done by `next`. This abstraction isn't adding much convenience. That would be ok if it was instead encapsulating a bunch of specialized knowledge, like the EcmaScript iteration protocol, which most simple uses of iteration and async iteration don't need to know about. But the `next` function here has to return an `IteratorResult`, which is visible only to someone who's already learned the iteration protocol.

Reply:
Thank you for this well-thought-out reply.  I have a pivot that I'll try out in this PR before I give up:

```js
/**
 * @template T
 * @callback AsyncGenerateYield mark the value to be `yield`ed to caller
 * @param {T} value the value to be yielded
 * @returns {T} the value
 */

/**
 * Return an async iterable that produces iterator results via `next`.
 *
 * @template T
 * @param {(yld: AsyncGenerateYield<T>) => T | Promise<T>} next produce a value
 * to yield or finish the iterator.  If `yld` is called, then the produced value
 * does not terminate the iterator.
 * @returns {AsyncIterable<T>}
 */
const asyncGenerate = ...;

/**
 * Return an async iterable that produces iterator results via next.
 *
 * @template T
 * @param {(done: AsyncGenerateDone<T>) => T | Promise<T>} next produce a value
 * to yield.  Once `done` is called, finish the iterator with the next produced
 * value
 * @returns {AsyncIterable<T>}
 */
```

Usage:

```js
  const gen = asyncGenerate(async yld => {
    const value = await Promise.resolve(nextValue);
    nextValue += 1;
    if (value < 4) {
      return yld(value);
    }
    return value;
  });
```

Reply:
I like the general idea. I'll iterate on the specifics after you push and let me know. But one thing immediately: What's the purpose of calling `Promise.resolve` above?

Also, what happens if `yld` is called "after" the promise for the result fulfills? Scare quotes on the "after" to indicate a possible race hazard.

Reply:
Okay, as discussed, the `asyncGenerate` function only exists to provide a way to create an `AsyncIterable` without using `Symbol`.  It has been reverted to the earlier implementation that minimally provides that (call `next()` to get a `{done,value}` result) and moved to `ring0`.